atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -316,35 +327,77 @@ private void notifyFailureListeners(final Throwable error) {
 
   /** Registers server handlers on the configured protocol. */
   private void registerHandlers(final RaftServerProtocol protocol) {
-    protocol.registerConfigureHandler(request -> runOnContext(() -> role.onConfigure(request)));
-    protocol.registerInstallHandler(request -> runOnContext(() -> role.onInstall(request)));
-    protocol.registerReconfigureHandler(request -> runOnContext(() -> role.onReconfigure(request)));
-    protocol.registerJoinHandler(request -> runOnContext(() -> role.onJoin(request)));
-    protocol.registerLeaveHandler(request -> runOnContext(() -> role.onLeave(request)));
-    protocol.registerTransferHandler(request -> runOnContext(() -> role.onTransfer(request)));
+    protocol.registerConfigureHandler(
+        request ->
+            handleRequestOnContext(
+                request, () -> role.onConfigure(request), ConfigureResponse::builder));
+    protocol.registerInstallHandler(
+        request ->
+            handleRequestOnContext(
+                request, () -> role.onInstall(request), InstallResponse::builder));
+    protocol.registerReconfigureHandler(
+        request ->
+            handleRequestOnContext(
+                request, () -> role.onReconfigure(request), ReconfigureResponse::builder));
+    protocol.registerJoinHandler(
+        request ->
+            handleRequestOnContext(request, () -> role.onJoin(request), JoinResponse::builder));
+    protocol.registerLeaveHandler(
+        request ->
+            handleRequestOnContext(request, () -> role.onLeave(request), LeaveResponse::builder));
+    protocol.registerTransferHandler(
+        request ->
+            handleRequestOnContext(
+                request, () -> role.onTransfer(request), TransferResponse::builder));
     protocol.registerAppendV1Handler(
-        request -> runOnContext(() -> role.onAppend(ProtocolVersionHandler.transform(request))));
+        request ->
+            handleRequestOnContext(
+                request,
+                () -> role.onAppend(ProtocolVersionHandler.transform(request)),
+                AppendResponse::builder));
     protocol.registerAppendV2Handler(
-        request -> runOnContext(() -> role.onAppend(ProtocolVersionHandler.transform(request))));
-    protocol.registerPollHandler(request -> runOnContext(() -> role.onPoll(request)));
-    protocol.registerVoteHandler(request -> runOnContext(() -> role.onVote(request)));
-  }
+        request ->
+            handleRequestOnContext(
+                request,
+                () -> role.onAppend(ProtocolVersionHandler.transform(request)),
+                AppendResponse::builder));
+    protocol.registerPollHandler(
+        request ->
+            handleRequestOnContext(request, () -> role.onPoll(request), PollResponse::builder));
+    protocol.registerVoteHandler(
+        request ->
+            handleRequestOnContext(request, () -> role.onVote(request), VoteResponse::builder));
+  }
+
+  private <T extends Builder<T, R>, R extends RaftResponse>
```

Comment:
:thought_balloon: I see what you mean. I agree it's a little weird in the context itself, and I think we all agree the context already does too much :smile: Registering the handlers might be behavior we want to extract its own class, it's well defined behavior. But I'm fine with leaving this here, as it makes sense together with the handler registration (or rather, it makes as much sense to have it here as it does the handler registration ;))

Reply:
I did try to take the handler registration out of `RaftContext`, but I didn't like the way I did it. :smile: 
One thing I noticed could be taken out is the implementation of `join`, `leave`, `transfer` etc. I will do it when I add force reconfigure.

Reply:
I will try out some refactoring to split up `RaftContext` and if it makes it better I will open a follow up PR.